File: backend/node/api/src/application/use-cases/update-pipeline.ts
import { inject } from "../../infra/DI/container";
import type { PipelineRepository } from "../../infra/repository/pipeline";

type Input = {
	id: string;
	name?: string;
	description?: string;
	canvas?: string;
};

export class UpdatePipelineUseCase {
	@inject("pipeline-repository")
	private declare readonly pipelineRepository: PipelineRepository;

	async execute(
		input: Input,
	): Promise<[undefined, undefined] | [undefined, Error]> {
		const [pipeline, error] = await this.pipelineRepository.getById(input.id);
		if (error) return [undefined, error];

		if (input.name !== undefined) {
			pipeline.setName(input.name);
		}

		if (input.description !== undefined) {
			pipeline.setDescription(input.description);
		}

		if (input.canvas !== undefined) {
			const [, canvasError] = pipeline.setCanvas(input.canvas);
			if (canvasError) return [undefined, canvasError];
		}

		const [, saveError] = await this.pipelineRepository.update(pipeline);
		if (saveError) return [undefined, saveError];

		return [undefined, undefined];
	}
}

File: backend/go/api/internal/pipeline/use-cases/update-pipeline.go
package usecases

import (
	"context"

	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/pipeline"
	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/shared/errors"
)

type UpdatePipelineUseCaseDependencies struct {
	PipelineRepository *pipeline.PipelineRepository
}

type UpdatePipelineUseCase struct {
	UpdatePipelineUseCaseDependencies
}

func NewUpdatePipelineUseCase(deps UpdatePipelineUseCaseDependencies) UpdatePipelineUseCase {
	return UpdatePipelineUseCase{
		UpdatePipelineUseCaseDependencies: deps,
	}
}

type UpdatePipelineInput struct {
	ID          string
	Name        *string
	Description *string
	Canvas      *string
}

func (u *UpdatePipelineUseCase) Execute(input UpdatePipelineInput) error {
	p, err := u.PipelineRepository.GetByID(context.Background(), input.ID)
	if err != nil {
		return err
	}

	if p == nil {
		return errors.PipelineNotFoundError(input.ID)
	}

	if input.Name != nil {
		p.SetName(input.Name)
	}

	if input.Description != nil {
		p.SetDescription(input.Description)
	}

	if input.Canvas != nil {
		if err := p.SetCanvas(input.Canvas); err != nil {
			return err
		}
	}

	return u.PipelineRepository.Update(context.Background(), p)
}
